### ISSUE TITLE
wsdd failing to start on FreeBSD

### ISSUE THREAD

**Issue by jasonemann** [christgau/wsdd#69]
wsdd 0.6.1 fails to start on FreeBSD when not passed any '-i' options (defaulting to all interfaces), or when provided with two '-i _interfacename_' options.  It will only start successfully if I specify interface IP addresses instead of names.

```
[root@filer ~]# uname -a
FreeBSD filer.home.arpa 12.2-RELEASE FreeBSD 12.2-RELEASE r366954 GENERIC  amd64

[root@filer ~]# ifconfig -a | egrep 'flags|inet |inet6 fe80|parent'
lan0: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> metric 0 mtu 1500
        inet 10.0.0.10 netmask 0xffffff00 broadcast 10.0.0.255
        inet6 fe80::21b:78ff:fe58:4ee6%lan0 prefixlen 64 scopeid 0x1
lan1: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> metric 0 mtu 1500
        inet6 fe80::21b:78ff:fe58:4ee7%lan1 prefixlen 64 scopeid 0x2
        inet 10.0.128.10 netmask 0xffffff00 broadcast 10.0.128.255
vlan64: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> metric 0 mtu 1500
        inet6 fe80::21b:78ff:fe58:4ee6%vlan64 prefixlen 64 scopeid 0x5
        inet 10.0.64.10 netmask 0xffffff00 broadcast 10.0.64.255
        vlan: 64 vlanpcp: 0 parent interface: lan0
```

lan0 and lan1 are renames of em0 and em1.  I only need wsdd to advertise on lan1 and vlan64.

With no '-i' options, it fails with:
```
[root@filer ~]# wsdd -v
2020-11-11 11:52:04,211:wsdd WARNING(pid 1527): no interface given, using all interfaces
2020-11-11 11:52:04,211:wsdd INFO(pid 1527): using pre-defined UUID 293a6c84-00bd-5f8c-8a60-2b21e9d138dd
2020-11-11 11:52:04,233:wsdd INFO(pid 1527): joined multicast group ('239.255.255.250', 3702) on 10.0.0.10%lan0
2020-11-11 11:52:04,246:wsdd INFO(pid 1527): joined multicast group ('ff02::c', 3702, 22364, 1) on fe80::21b:78ff:fe58:4ee6%lan0
2020-11-11 11:52:04,249:wsdd INFO(pid 1527): joined multicast group ('ff02::c', 3702, 22364, 2) on fe80::21b:78ff:fe58:4ee7%lan1
2020-11-11 11:52:04,252:wsdd INFO(pid 1527): joined multicast group ('239.255.255.250', 3702) on 10.0.128.10%lan1
Traceback (most recent call last):
  File "/usr/local/bin/wsdd", line 1650, in <module>
    sys.exit(main())
  File "/usr/local/bin/wsdd", line 1597, in main
    nm = RouteSocketAddressMonitor(s)
  File "/usr/local/bin/wsdd", line 923, in __call__
    obj.enumerate()
  File "/usr/local/bin/wsdd", line 1275, in enumerate
    self.parse_route_socket_response(rt_buf.raw, True)
  File "/usr/local/bin/wsdd", line 1311, in parse_route_socket_response
    intf, addr_mask, rtm_type, intf_flags)
  File "/usr/local/bin/wsdd", line 1360, in parse_addrs
    self.handle_new_address(addr, addr_family, intf)
  File "/usr/local/bin/wsdd", line 1016, in handle_new_address
    mch, WSDHttpRequestHandler, mch.family, self.selector))
  File "/usr/local/bin/wsdd", line 799, in __init__
    super().__init__(mch.listen_address, RequestHandlerClass)
  File "/usr/local/lib/python3.7/socketserver.py", line 452, in __init__
    self.server_bind()
  File "/usr/local/bin/wsdd", line 805, in server_bind
    super().server_bind()
  File "/usr/local/lib/python3.7/http/server.py", line 137, in server_bind
    socketserver.TCPServer.server_bind(self)
  File "/usr/local/lib/python3.7/socketserver.py", line 466, in server_bind
    self.socket.bind(self.server_address)
socket.gaierror: [Errno 8] Name does not resolve
```

With '-i vlan64 -i lan1', it fails with:
```
[root@filer ~]# wsdd -v -i vlan64 -i lan1
2020-11-11 11:52:57,401:wsdd INFO(pid 1530): using pre-defined UUID 293a6c84-00bd-5f8c-8a60-2b21e9d138dd
2020-11-11 11:52:57,423:wsdd INFO(pid 1530): joined multicast group ('ff02::c', 3702, 22364, 2) on fe80::21b:78ff:fe58:4ee7%lan1
2020-11-11 11:52:57,434:wsdd INFO(pid 1530): joined multicast group ('239.255.255.250', 3702) on 10.0.128.10%lan1
Traceback (most recent call last):
  File "/usr/local/bin/wsdd", line 1650, in <module>
    sys.exit(main())
  File "/usr/local/bin/wsdd", line 1597, in main
    nm = RouteSocketAddressMonitor(s)
  File "/usr/local/bin/wsdd", line 923, in __call__
    obj.enumerate()
  File "/usr/local/bin/wsdd", line 1275, in enumerate
    self.parse_route_socket_response(rt_buf.raw, True)
  File "/usr/local/bin/wsdd", line 1311, in parse_route_socket_response
    intf, addr_mask, rtm_type, intf_flags)
  File "/usr/local/bin/wsdd", line 1360, in parse_addrs
    self.handle_new_address(addr, addr_family, intf)
  File "/usr/local/bin/wsdd", line 1016, in handle_new_address
    mch, WSDHttpRequestHandler, mch.family, self.selector))
  File "/usr/local/bin/wsdd", line 799, in __init__
    super().__init__(mch.listen_address, RequestHandlerClass)
  File "/usr/local/lib/python3.7/socketserver.py", line 452, in __init__
    self.server_bind()
  File "/usr/local/bin/wsdd", line 805, in server_bind
    super().server_bind()
  File "/usr/local/lib/python3.7/http/server.py", line 137, in server_bind
    socketserver.TCPServer.server_bind(self)
  File "/usr/local/lib/python3.7/socketserver.py", line 466, in server_bind
    self.socket.bind(self.server_address)
socket.gaierror: [Errno 8] Name does not resolve
```

With IP addresses used instead of interface names, it starts OK:
```
[root@filer ~]# wsdd -v -i 10.0.64.10 -i 10.0.128.10
2020-11-11 11:53:56,441:wsdd INFO(pid 1533): using pre-defined UUID 293a6c84-00bd-5f8c-8a60-2b21e9d138dd
2020-11-11 11:53:56,462:wsdd INFO(pid 1533): joined multicast group ('239.255.255.250', 3702) on 10.0.128.10%lan1
2020-11-11 11:53:56,476:wsdd INFO(pid 1533): joined multicast group ('239.255.255.250', 3702) on 10.0.64.10%vlan64
```

**Comment by jasonemann** [comment]
Another example, with vlan interface replaced by another physical interface:
```
[root@filer ~]# ifconfig -a | egrep 'flags|inet |inet6 fe80'
lan1: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> metric 0 mtu 1500
        inet 10.0.16.10 netmask 0xffffff00 broadcast 10.0.16.255
lan2: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> metric 0 mtu 1500
        inet6 fe80::21b:78ff:fe58:4ee7%lan2 prefixlen 64 scopeid 0x2
        inet 10.0.128.10 netmask 0xffffff00 broadcast 10.0.128.255
lan0: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> metric 0 mtu 1500
        inet 10.0.0.10 netmask 0xffffff00 broadcast 10.0.0.255
        inet6 fe80::3e4a:92ff:fe77:c459%lan0 prefixlen 64 scopeid 0x3

[root@filer ~]# wsdd -v
2020-11-14 09:11:16,502:wsdd WARNING(pid 3650): no interface given, using all interfaces
2020-11-14 09:11:16,503:wsdd INFO(pid 3650): using pre-defined UUID 293a6c84-00bd-5f8c-8a60-2b21e9d138dd
2020-11-14 09:11:16,525:wsdd INFO(pid 3650): joined multicast group ('239.255.255.250', 3702) on 10.0.16.10%lan1
2020-11-14 09:11:16,538:wsdd INFO(pid 3650): joined multicast group ('ff02::c', 3702, 22364, 2) on fe80::21b:78ff:fe58:4ee7%lan2
2020-11-14 09:11:16,541:wsdd INFO(pid 3650): joined multicast group ('239.255.255.250', 3702) on 10.0.128.10%lan2
Traceback (most recent call last):
  File "/usr/local/bin/wsdd", line 1650, in <module>
    sys.exit(main())
  File "/usr/local/bin/wsdd", line 1597, in main
    nm = RouteSocketAddressMonitor(s)
  File "/usr/local/bin/wsdd", line 923, in __call__
    obj.enumerate()
  File "/usr/local/bin/wsdd", line 1275, in enumerate
    self.parse_route_socket_response(rt_buf.raw, True)
  File "/usr/local/bin/wsdd", line 1311, in parse_route_socket_response
    intf, addr_mask, rtm_type, intf_flags)
  File "/usr/local/bin/wsdd", line 1360, in parse_addrs
    self.handle_new_address(addr, addr_family, intf)
  File "/usr/local/bin/wsdd", line 1016, in handle_new_address
    mch, WSDHttpRequestHandler, mch.family, self.selector))
  File "/usr/local/bin/wsdd", line 799, in __init__
    super().__init__(mch.listen_address, RequestHandlerClass)
  File "/usr/local/lib/python3.7/socketserver.py", line 452, in __init__
    self.server_bind()
  File "/usr/local/bin/wsdd", line 805, in server_bind
    super().server_bind()
  File "/usr/local/lib/python3.7/http/server.py", line 137, in server_bind
    socketserver.TCPServer.server_bind(self)
  File "/usr/local/lib/python3.7/socketserver.py", line 466, in server_bind
    self.socket.bind(self.server_address)
socket.gaierror: [Errno 8] Name does not resolve

[root@filer ~]# wsdd -v -i lan1 -i lan2
2020-11-14 09:11:55,958:wsdd INFO(pid 3653): using pre-defined UUID 293a6c84-00bd-5f8c-8a60-2b21e9d138dd
2020-11-14 09:11:55,980:wsdd INFO(pid 3653): joined multicast group ('239.255.255.250', 3702) on 10.0.16.10%lan1
2020-11-14 09:11:55,993:wsdd INFO(pid 3653): joined multicast group ('ff02::c', 3702, 22364, 2) on fe80::21b:78ff:fe58:4ee7%lan2
2020-11-14 09:11:55,996:wsdd INFO(pid 3653): joined multicast group ('239.255.255.250', 3702) on 10.0.128.10%lan2
Traceback (most recent call last):
  File "/usr/local/bin/wsdd", line 1650, in <module>
    sys.exit(main())
  File "/usr/local/bin/wsdd", line 1597, in main
    nm = RouteSocketAddressMonitor(s)
  File "/usr/local/bin/wsdd", line 923, in __call__
    obj.enumerate()
  File "/usr/local/bin/wsdd", line 1275, in enumerate
    self.parse_route_socket_response(rt_buf.raw, True)
  File "/usr/local/bin/wsdd", line 1311, in parse_route_socket_response
    intf, addr_mask, rtm_type, intf_flags)
  File "/usr/local/bin/wsdd", line 1360, in parse_addrs
    self.handle_new_address(addr, addr_family, intf)
  File "/usr/local/bin/wsdd", line 1016, in handle_new_address
    mch, WSDHttpRequestHandler, mch.family, self.selector))
  File "/usr/local/bin/wsdd", line 799, in __init__
    super().__init__(mch.listen_address, RequestHandlerClass)
  File "/usr/local/lib/python3.7/socketserver.py", line 452, in __init__
    self.server_bind()
  File "/usr/local/bin/wsdd", line 805, in server_bind
    super().server_bind()
  File "/usr/local/lib/python3.7/http/server.py", line 137, in server_bind
    socketserver.TCPServer.server_bind(self)
  File "/usr/local/lib/python3.7/socketserver.py", line 466, in server_bind
    self.socket.bind(self.server_address)
socket.gaierror: [Errno 8] Name does not resolve

[root@filer ~]# wsdd -v -i 10.0.16.10 -i 10.0.128.10
2020-11-14 09:14:52,520:wsdd INFO(pid 3660): using pre-defined UUID 293a6c84-00bd-5f8c-8a60-2b21e9d138dd
2020-11-14 09:14:52,541:wsdd INFO(pid 3660): joined multicast group ('239.255.255.250', 3702) on 10.0.16.10%lan1
2020-11-14 09:14:52,555:wsdd INFO(pid 3660): joined multicast group ('239.255.255.250', 3702) on 10.0.128.10%lan2
```

**Comment by christgau** [comment]
TL;DR: Possible Workaround: could you try to add the IP of the vlans to `/etc/hosts` or to your DNS (if available)?

Based on the error code and after digging into the Python library sources it looks_to me that Python tries to resolve the address, again into... an address. And the reason for this might be that wsdd allocates a string buffer to parse the response from the route socket and extracts the (binary) ip address from it. Thus, Python assumes it to be a string. I have a small fix in mind, but I'm currently out of time to give it a try. Could you try the workaround in the meantime?

**Comment by jasonemann** [comment]
Thanks for your response.

I've tried as you suggested, but no difference in behaviour:

```
[root@filer ~]# tail -7 /etc/hosts
10.0.0.10       filer-mgt.home.arpa
fe80::3e4a:92ff:fe77:c459       filer-mgt.home.arpa

10.0.16.10      filer.home.arpa

10.0.128.10     filer.lab.home.arpa
fe80::21b:78ff:fe58:4ee7        filer.lab.home.arpa

[root@filer ~]# wsdd -v
2020-11-28 14:11:00,612:wsdd WARNING(pid 13847): no interface given, using all interfaces
2020-11-28 14:11:00,612:wsdd INFO(pid 13847): using pre-defined UUID 293a6c84-00bd-5f8c-8a60-2b21e9d138dd
2020-11-28 14:11:00,635:wsdd INFO(pid 13847): joined multicast group ('239.255.255.250', 3702) on 10.0.16.10%lan1
2020-11-28 14:11:00,645:wsdd INFO(pid 13847): joined multicast group ('ff02::c', 3702, 22364, 2) on fe80::21b:78ff:fe58:4ee7%lan2
2020-11-28 14:11:00,648:wsdd INFO(pid 13847): joined multicast group ('239.255.255.250', 3702) on 10.0.128.10%lan2
Traceback (most recent call last):
  File "/usr/local/bin/wsdd", line 1650, in <module>
    sys.exit(main())
  File "/usr/local/bin/wsdd", line 1597, in main
    nm = RouteSocketAddressMonitor(s)
  File "/usr/local/bin/wsdd", line 923, in __call__
    obj.enumerate()
  File "/usr/local/bin/wsdd", line 1275, in enumerate
    self.parse_route_socket_response(rt_buf.raw, True)
  File "/usr/local/bin/wsdd", line 1311, in parse_route_socket_response
    intf, addr_mask, rtm_type, intf_flags)
  File "/usr/local/bin/wsdd", line 1360, in parse_addrs
    self.handle_new_address(addr, addr_family, intf)
  File "/usr/local/bin/wsdd", line 1016, in handle_new_address
    mch, WSDHttpRequestHandler, mch.family, self.selector))
  File "/usr/local/bin/wsdd", line 799, in __init__
    super().__init__(mch.listen_address, RequestHandlerClass)
  File "/usr/local/lib/python3.7/socketserver.py", line 452, in __init__
    self.server_bind()
  File "/usr/local/bin/wsdd", line 805, in server_bind
    super().server_bind()
  File "/usr/local/lib/python3.7/http/server.py", line 137, in server_bind
    socketserver.TCPServer.server_bind(self)
  File "/usr/local/lib/python3.7/socketserver.py", line 466, in server_bind
    self.socket.bind(self.server_address)
socket.gaierror: [Errno 8] Name does not resolve
```

**Comment by christgau** [comment]
Too bad. I'll keep you informed if I have an update on this issue but it could take some time.

**Comment by christgau** [comment]
It took a while but I was able to reproduce the issue with v0.6.1. However, a8fd150d fixed the issue. I think it's time for a new release. Meanwhile you could try the master branch.
